### PR TITLE
reduce cache retry load

### PR DESCRIPTION
**Pull request by fspmarshall**
Introduces a new exponential backoff type and applies it to non-control-plane caches. The intent of this change is to help us better reduce thundering herd effects in large clusters (10k+ agents).

The new exponential backoff starts off with comparatively sized delays, but escalates much quicker on recurring errors. We also increase the maximum retry for non-control-plane caches from 90s to 256s, though most users will actually experience an increase from 60s to 256s since the switch from 60s to 90s was very recent.

The old standard retry behavior resulted in an approximate sequence of `0s, 12s, 24s, 36s, 48s, 60s` for all instances.  The new behavior as of this change results in an approximate sequence of `0s, 5s, 10s, 20s, 40s, 80s, 90s` for control-plane elements and an approximate sequence of `0s, 16s, 32s, 64s, 128s, 256s` for peripheral agents.

---

In addition to the primary behavioral changes described above, this PR also makes two more minor changes.  Caches now allow their full max backoff for the init event to show up instead of a fixed one minute (slow init event propagation has been observed as a source of excessive cache re-init errors in high load clusters).  The maximum backoff used by the cache is now also configurable like so:

```yaml
teleport:
  cache:
    enabled: yes
    max_backoff: 12m
```

While the default value of 256s is probably sufficiently large for most clusters, _very_ large clusters might benefit from bumping this even higher.

### CODE REVIEW

File: lib/defaults/defaults.go [hunk 1]
```diff
@@ -308,6 +308,10 @@ const (
 	// MaxWatcherBackoff is the maximum retry time a watcher should use in
 	// the event of connection issues
 	MaxWatcherBackoff = 90 * time.Second
+
+	// MaxLongWatcherBackoff if the maximum backoff used for watchers that incur high cluster-level
```
Comment (espadolini) [inline lib/defaults/defaults.go:312]:
```suggestion
	// MaxLongWatcherBackoff is the maximum backoff used for watchers that incur high cluster-level
```

File: api/utils/retryutils/retryv2.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+//
+// NOTE: drivers must be stateless. If a stateful driver needs to be implemented in the
+// future, this interface will need to be extended to support safe use of Retry.Clone.
+type Driver interface {
+	// Duration calculates the step-specific delay for a given attempt. Excludes
+	// base duration and jitter, which are applied by the outer retry instance.
+	Duration(attempt int64) time.Duration
+
+	// Check verifies the correctness of any driver-internal parameters.
+	Check() error
```
Comment (espadolini) [inline api/utils/retryutils/retryv2.go:42]: Should this be part of the interface, or should the constructor be fallible?

File: api/utils/retryutils/retryv2.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+// NewLinearDriver creates a linear retry driver with the supplied step value. Resulting retries
+// have increase their backoff by a fixed step amount on each increment, with the first retry
+// having a base step amount of zero.
+func NewLinearDriver(step time.Duration) Driver {
+	return &linearDriver{step}
```
Comment (espadolini) [inline api/utils/retryutils/retryv2.go:49]: Both `linearDriver` and `exponentialDriver` can (should?) implement `Driver` by value.

File: api/utils/retryutils/retryv2.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	if r.Jitter != nil {
+		a = r.Jitter(a)
+	}
+
+	return a
+}
+
+// After returns channel that fires with timeout
+// defined in Duration method, as a special case
+// if Duration is 0 returns a closed channel
```
Comment (espadolini) [inline api/utils/retryutils/retryv2.go:229]: We don't rely on the special case anywhere for `retryutils.Linear`, can we get rid of it?

**Comment by zmb3** [comment]
Wouldn't hurt to rerun the GHA tests a few extra times on this, as these types of changes have caused flakiness in the past.

File: api/utils/retryutils/retryv2.go [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	First time.Duration
+	// Driver generates the underlying progression of delays. Cannot be nil.
+	Driver Driver
+	// Max is a maximum value of the progression,
+	// can't be 0
+	Max time.Duration
+	// Jitter is an optional jitter function to be applied
+	// to the delay.  Note that supplying a jitter means that
+	// successive calls to Duration may return different results.
+	Jitter Jitter `json:"-"`
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/utils/retryutils/retryv2.go:142]: Apparently the old `Linear` was sometimes marshaled to json as part of logging, and the `Jitter` broke it.  IDK if it's still a concern, but seems harmless to keep around.

File: api/utils/retryutils/retryv2.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+		return 0
+	}
+
+	if a > r.Max {
+		a = r.Max
+	}
+
+	if r.Jitter != nil {
+		a = r.Jitter(a)
+	}
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply api/utils/retryutils/retryv2.go:229]: All jitters are subtractive, so by applying `Max` first, we ensure that the post-jitter value is less than or equal to `Max`.

File: api/utils/retryutils/retryv2.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+type exponentialDriver struct {
+	base time.Duration
+}
+
+func (d exponentialDriver) Duration(attempt int64) time.Duration {
+	if attempt > 16 {
+		// 16 will exceed any reasonable Max value already, and we don't
+		// want to accidentally wrap and end up w/ negative durations.
+		attempt = 16
+	}
```
Comment (espadolini) [inline api/utils/retryutils/retryv2.go:99]:
```suggestion
	if attempt > maxAttempts {
		attempt = maxAttempts
	}
```

**Review by espadolini — APPROVED** [review]

**Review by gabrielcorado — APPROVED** [review]

**Pull request merged**